Feature/parallel ssro

nathlacroix opened the pull request:
Addresses #70.

Changes proposed in this pull request:
- Parallel SSRO calib measurement: **mqm.measure_ssro()**. uses calibration points to create SSRO sequences with an arbitrary number of qubits in parallel. Also more flexible than previous single qubit measurement in terms of which states can be measured. See doc string for more. Basic functionalities tested on BF1.
-  Parallel SSRO calib analysis. Flexible in terms of which qubits to analyze. Note that **90% of analysis is time required to plot and save the figures**.
- Parallel optimal weights calibration: **mqm.find_optimal_weights()**. Can measure an arbitrary number of qubits in parallel as long as they are not on the same UHF. Tested on virtual setup and on single qubit calib on BF1 (all qubits are on same UHF and therefore the multi qubit version cannot be tested)
- parallel optimal weights Analysis: we were waiting for this analysis since 2017 ;). A basic version of it is now implemented in timedomain_analysis. 

Inviting @stephlazar to review. FYI @christiankraglundandersen @chellings @antsr

nathlacroix committed:
Add possibility to give segments when creating Sequence

In the same spirit as for the Segment, segments can be given already when instantiating the Sequence object.

nathlacroix committed:
Fix timing readout in cal_points segments

The timing of the readout pulse wasn't coming after the last pulse (in
time unit) but after the end of the last pulse of the pulse list.
This typically results in errors when some qubits are prepared in the e
state and others in the f state.
Corrected by using the multireferencing such that the readout pulses
would start after the last state preparation pulse

nathlacroix committed:
Add parallel SSRO measurement

Analysis to come.

nathlacroix committed:
Update Parallel SSRO measurement

- allow for custom state sequence
- extend doc string
- rename parameters

nathlacroix committed:
Minor changes for Single Qubit SSRO analysis

- include additional labels
- aesthetics

nathlacroix committed:
Add parallel SSRO analysis

- based on MultiqubitTimedomain: while this seems strange at first,
the required "per qubit" handling of the data is quite similar.
I could not include the analysis in readout_analysis because it would
create a circular dependency between timedomain_analysis.py and readout_analysis.py
When migrating to analysis_v3, one should be able to just reuse common nodes.
For now, it resulted in less code duplication to do it in that way.
- Reuses all static functions from the SSRO Qutrit single shot analysis
- Basic functionalities implemented and tested on virtual setup

nathlacroix committed:
Reduce code duplication in MultiQutritSSRO analyis

-Move the NCC class such that MultiQutrit analysis can access it.

nathlacroix committed:
Fix encoding when saving to hdf5

keys for attributes must be of string type. If it is not the case (e.g. the key of dictionary is a integer) then this results in an Exception and corresponding key/value pair is not saved.
To fix this, a representation of the key is created. To retrieve the original value, "eval"
must be used on the representation, i.e. eval(repr(key)) == key

nathlacroix committed:
Show codeword mapping

Display the codeword mapping on plot with thresholds for SSRO analysis

nathlacroix committed:
Add parallel SSRO analysis after measurement

nathlacroix committed:
Add warning no data left after preselection

nathlacroix committed:
Update docstring parallel SSRO

nathlacroix committed:
Fix minor SSRO related bugs/features

- change meaning of preselection flag: if None, then only prep_params is used
- prepare qubits before creating channel map
- raise error if # states != # qubits

nathlacroix committed:
Fix multi timestamp metadata when using get_param_value

- the metadata in readout dict wasn't properly updated to a dict

nathlacroix committed:
Allow active reset for parallel SSRO

nathlacroix committed:
Add parallel weights measurement for SSRO (WIP)

nathlacroix committed:
Save data_file in BaseAnalysis

nathlacroix committed:
Add Multi qutrit timetrace analysis to compute optimal weights

nathlacroix committed:
Modify Timetrace analysis

- add doc strings
- add option to scale weights

nathlacroix committed:
Modify timetraceAnalysis

- add possibility to not specify qb names
- correct bug second pair of weights not plotted

nathlacroix committed:
Add timeTrace analysis to measurement

nathlacroix committed:
Merge branch 'qudev_master' into feature/parallel_SSRO

# Conflicts:
#	pycqed/analysis_v2/timedomain_analysis.py
#	pycqed/measurement/calibration_points.py

nathlacroix committed:
Clean up merge with qudev_master

- add preselection datafilter
- add default projected datadict value as dict

nathlacroix committed:
Remove return sequence

nathlacroix committed:
Remove comment in SSRO analysis

stephlazar commented on pycqed/measurement/multi_qubit_module.py:
```diff
@@ -447,6 +449,256 @@ def measure_multiplexed_readout(qubits, liveplot=False,
         ))
 
+def measure_ssro(qubits, states=('g', 'e'), n_shots=10000, label=None,
+                 preselection=True, all_states_combinations=False, upload=True,
```
Shouldn't measure_ssro and measure_optimized weights use the Device object?

nathlacroix replied:
It could but I'm not sure I would understand for what exactly it would be used as those measurement function do not use anything related to two-qubit gates. To my understanding, this was the main use-case of the device object so far (?). Or did we agree that all functions should use the device object no matter whether it uses it or not? I could see in the future that we for instance add a config to the device object that maybe specifies which qubit can be calibrated simultaneously etc., which might be relevant when we start automating calibration etc. but for now I guess the user in the notebook is still taking care of that. In any case, whether we do it now or later, it will be simple to just add the device object as a parameter to this function

stephlazar replied:
As i understood it, we would like to use the device object for all multi-qubit algorithms. I also added it to the multi-qubit time domain as per Ants' suggestion. Otherwise we create a mess if only some functions use it :p 
It has wrappers around get_operation_dict and get_prep_params which should become the new standard to use for all multi-qubit measurements.

chellings replied:
And the device object stores all qubits objects. I agreed with Michael that the measure functions should then accept qubit names as well as qubits objects:
`qubits = [dev.get_qb(qb) if isinstance(qb, str) else qb for qb in qubits]`

I think there is no need to modify old functions if they do not use 2qb gates. But for new multi-qubit functions, it might make sense to already stick to this concept, even if they do not need 2qb gates. What do you think?

nathlacroix replied:
Ok for me, I will then add this now.

stephlazar commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -252,7 +252,7 @@ def get_data_from_timestamp_list(self):
             h5mode = self.options_dict.get('h5mode', 'r+')
             h5filepath = a_tools.measurement_filename(folder)
-            data_file = h5py.File(h5filepath, h5mode)
+            self.data_file = h5py.File(h5filepath, h5mode)
```
I removed making data_file an attribute. This is useless because we close it at the end of this function, so you even if you can call class_instance.data_file, this will just give you a closed file. Why did you make it an attribute?

nathlacroix replied:
It is closed only if an exception occurs, which I think is good. Otherwise, I think we should close it only at the end of the analysis (see e.g. line 5682 in timedomain analysis, but in fact I agree there it could also be in a try-except clause). 
To me being able to access the data_file during the analysis is useful to have access to anything in the hdf5 file and I think it should be possible in an easy way without having to reconstruct it with h5py. I for instance use it at line 5784 in time domain analysis.

If you think it is better not to make it an attribute, then I would suggest to create a wrapper of ba.get_hdf_param_value() that is not a static function, and that would create the required data_file object from self and then pass the object to get_hdf_param_value(). Do you think this would be better? In that way, whenever you want (also in the notebook), you could just do e.g. analysis_obj.get_hdf_value("Instrument settings/qb1", "T1") to get the value without needing to import h5py and create an data_file object. The disadvantage of this feature is that if you use it in a loop then it might slow you down a bit as you open and close the file each time, compared to having it as an attribute.

stephlazar replied:
Ok I see why you need it. However, you should use the params_dict and numeric_params for your use case in line 5784. That's how analysis_v2 was intended to be used, everything passed to params_dict and numeric_params are extracted and saved in raw_data_dict. See the init or RabiAnalysis for example of how to use this. I think this is nice because then we don't need access to the data_file throughout the methods. I think you should do it like this as well and remove the attribute self.data_file.

nathlacroix replied:
Yes, I thought about using that method indeed. I decided not to because it relies on passing the qubit names, and I think that those should not be mandatory if they can be inferred from the channel map , or calibration points, etc. i.e. the default should be to analyze all qubits in the file. However, you can get those qb names only after extracting the data, so not in the init but after extract_data(). On the other hand, the numeric params have to be done in the init :D, so one cannot have both. 
But okay, it is not a critical thing to mandatorily pass the qubit_names, so I will use that if it makes more sense. I might still also create the other function I spoke about, as I think it could be useful even in the notebook to access the hdf5 file

stephlazar replied:
I'm not sure I understand. Looking at your class, you pass the qb_names in the init so you have access to them there, and that's where you create the params_dict/numeric_params (as is done in RabiAnalysis). Since prepare_plots is one of the last things you call, these params will already have been extracted and are in raw_data_dict. In RabiAnalysis I do exactly this: create params_dict in init, then just retrieve them in prepare_plots. And it looks to me like your use case is exactly the same.

stephlazar replied:
I realized now that your were talking in general, that it would be nice to not have to pass qb_names.

nathlacroix replied:
Yes, I think I have now implemented something that respects both our ideas. It uses the numeric param by default but if qb_names is not specified then it uses the wrapper I have created. So in any case, the self.data_file isn't there anymore

nathlacroix replied:
what do you think? (see commit [521dd3e ](https://github.com/QudevETH/PycQED_py3/pull/79/commits/521dd3e1e1245a3733c097e21d55e26596687dd5) )

stephlazar commented on pycqed/analysis_v2/timedomain_analysis.py:
```diff
@@ -5627,4 +5639,591 @@ def prepare_plots(self):
\ No newline at end of file
+                 'legend_pos': 'upper right'})
+
+class MultiQutrit_Timetrace_Analysis(ba.BaseDataAnalysis):
```
Why is this not in readout_analysis?

nathlacroix replied:
I could move it there, no strong opinion

nathlacroix replied:
Though I cannot do it for the multiqutrit ssro analysis because it is based on multiqubit timedomain (a bit unfortunate from a semantic point of view I admit but I would otherwise duplicate big chuncks of code) and importing tda in readout analysis would create a circular import (readout analysis is already imported in timedomain analysis)

stephlazar replied:
I see. I don't have a strong preference, we can leave them in tda. I just wasn't sure why you chose to implement them there.

stephlazar reviewed (changes requested):
I did not look in detail at the changes in readout_analysis.py since I am not very familiar with that file, but I trust that you did the right thing :)

nathlacroix committed:
Remove self.data_file

- removes attribute self.data_file in base analysis
- adds a wrapper function to easily access the datafile from the analysis object

stephlazar commented on pycqed/analysis_v2/base_analysis.py:
```diff
@@ -230,6 +230,30 @@ def get_hdf_param_value(group, param_name):
+        h5mode = self.options_dict.get('h5mode', 'r+')
+        folder = a_tools.get_folder(self.timestamps[hdf_file_index])
+        h5filepath = a_tools.measurement_filename(folder)
+        data_file = h5py.File(h5filepath, h5mode)
```
You should close the data_file even when there was not exception.

stephlazar reviewed (changes requested):
Looks good otherwise! Thanks for the change :)

nathlacroix committed:
Close data_file after usage

nathlacroix committed:
Add device object to measure_ssro and find_optimal_weights

- add device object in signature
- add get_qubit() in device objects, as a wrapper to call qubit names and qubit objects
- add possibility to restrict the operations dict to a subset of qubits

nathlacroix commented:
@stephlazar added a final commit addressing your previous comment related to the device object. Also created a wrapper function in the device object to address @chellings ' comment about accepting strings and qubit objects, to avoid having that line of code everywhere in our code base.

nathlacroix reviewed (commented):
changed the base branch to qudev_master (but it is identical to S17, so no need to redo review)
